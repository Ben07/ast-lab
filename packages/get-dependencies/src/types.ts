import { ALL_EXPORTS } from './constants';

export type Module = string;
type Imported = string;
type ImportedAlias = string;
export type Exported = string;

export type ModuleExported = Set<Exported> | null | typeof ALL_EXPORTS;
export type ModuleImported = Map<Imported, {
  alias: ImportedAlias | null,
  affectedExports: ModuleExported
}>;

export type Dependents = Map<Module, ModuleImported>;
export type Exports = Map<Module, ModuleExported>;

export type Loader = (absoluteFilePath: string) => Promise<string>;
export type Resolver = (mod: string, source: string) => Promise<string | void>;

export type PathNode = {
  source: Module | null,
  importModule: Module,
  i2e: ModuleImported | null,
  prev: PathNode | null
};

export type Options = {
  loader?: Loader,
  extensions: string[],
  resolver: Resolver
};

export type VisitedNode = PathNode[];
export type Visited = Map<Module, VisitedNode>;
