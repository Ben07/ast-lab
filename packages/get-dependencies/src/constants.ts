export const ALL_MODULES = '*';
export const DEFAULT = 'default';
export const ALL_EXPORTS: true = true;
export const DEFAULT_EXTENSIONS = ['.js', '.ts', '.jsx', '.tsx'];