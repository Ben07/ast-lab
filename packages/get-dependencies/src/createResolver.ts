import enhancedResolve, { ResolverFactory } from 'enhanced-resolve';
import path from 'path';

import isCore from './isCore';
import isThirdParty from './isThirdParty';

export default function createResolver(
  opts: ResolverFactory.ResolverOption = {}
) {
  return async (mod: string, source: string = '/'): Promise<string | void> => {
    if (!mod || isCore(mod)) {
      return;
    }
    if (path.isAbsolute(mod)) {
      return mod;
    }
    const basedir = path.dirname(source || '/');
    const result = await new Promise((resolve, reject) => 
      enhancedResolve.create(opts)(
        basedir, 
        mod, 
        (err: Error, res: string | undefined) => {
          if (err) {
            return reject(err.message);
          }
          if (!res) {
            console.log(`Couldn't find module ${mod}!`);
            return resolve();
          }
          if (isThirdParty(res)) {
            return resolve();
          }
          return resolve(res);
        }
      )
    ) as string | void;
    return result;
  }
}
