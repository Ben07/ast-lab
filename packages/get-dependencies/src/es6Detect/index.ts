// @ts-ignore
import Walker from 'node-source-walk';
import { Node, ModuleSpecifier, Program } from 'estree';
import { Dependents, ModuleImported } from '../types';
// import getPatternIdentifiers from './getPatternIdentifiers';
import fs from 'fs';
import _importSpecifier2Dependents from './importSpecifier2Dependents';
import depsFromExportNamed from './depsFromExportNamed';
import _depsFromExportAll from './depsFromExportAll';

type Options = {
  inDetail?: boolean,
  resolve: (mod: string) => string | void,
  load: (file: string) => string
}

/**
 * Get ES6 file dependencies (module and imported defination)
 * @todo support import affected export mapping
 * @param src {string} file content
 * @param [inDetail] {boolean} NOT FULLY SUPPORTED. Get affected exports
 * @return {Map<string, Set<name> | null>}
 */
export default function getEs6Dependents(
  src: string,
  {
    inDetail,
    resolve,
    load = (file: string) => fs.readFileSync(file, 'utf8')
  }: Options): Dependents {
  const walkerIns = new Walker();
  let dependencies: Dependents = new Map();
  src = load(src);
  if (src === '') {
    return dependencies;
  }
  if (typeof src === 'undefined') {
    throw new Error('src is undefined!');
  }

  const ast: Program = walkerIns.parse(src).program;
  const importSpecifier2Dependents = _importSpecifier2Dependents(inDetail);
  const depsFromExportAll = _depsFromExportAll({ resolve, load });
  const rootDeclarations: string[] = ast.body.reduce((roots: string[], node: Node) => {
    switch (node.type) {
      case 'ImportDeclaration':{
        const modulePath = node.source && node.source.value as string;
        if (!modulePath) { return roots; }
        if (!dependencies.has(modulePath)) {
          dependencies.set(modulePath, new Map() as ModuleImported);
        }
        node.specifiers.forEach((specifier: ModuleSpecifier) => {
          const depMap = dependencies.get(modulePath) as ModuleImported;
          dependencies.set(modulePath, importSpecifier2Dependents(depMap, specifier));
          roots.push(specifier.local.name);
        });
        break;
      }
      case 'ExportNamedDeclaration':{
        const modulePath = node.source && (node.source.value as string);
        if (!modulePath) { return roots; }
        dependencies = depsFromExportNamed(dependencies, node);
        (dependencies.get(modulePath) as ModuleImported).forEach(
          ({ alias }, name) => roots.push(alias || name)
        );
        break;
      }
      case 'ExportAllDeclaration':
        dependencies = depsFromExportAll(dependencies, node);
        /** @todo find roots for export all */
        break;
      case "ExportDefaultDeclaration":
        roots.push('default');
        break;
      // case 'VariableDeclaration':
      //   node.declarations.forEach(({ id }) => {
      //     roots.concat(getPatternIdentifiers(id));
      //   });
      //   break;
      // case 'FunctionDeclaration':
      // case 'ClassDeclaration':
      //   roots.concat(getPatternIdentifiers(node.id));
      //   break;
    }
    return roots;
  }, [] as string[]);
  console.log(rootDeclarations);

  /** @todo support dynamic import */
  // walkerIns.walk(ast, function(node: Node) {
  //   switch (node.type) {
  //     case 'CallExpression':
  //       if (['require', 'import'].includes((node.init.callee as Identifier).name)) {
  //         const { id, arguments: args } = node;
  //         if (id.type === 'ObjectPattern') {

  //         } else {
  //           dependencies.set(node.source.value, null);
  //         }
  //       }
  //       return;
  //     case 'Identifier':

  //       break;
  //   }
  // });

  return dependencies;
}
