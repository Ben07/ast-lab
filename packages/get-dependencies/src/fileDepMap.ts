import fs from 'fs';
import ecmaStats from 'es-stats';
import { Module, Member, Entry, Import, DependencyMap, AffectedMap, Options } from 'ast-lab-types';
import _debug from 'debug';
import { MODULE_ALL } from './constants';

const debug = _debug('get-dependencies:file');

function updateDependencyMap(depMap: DependencyMap, mod: Module, member: Member, entry: Entry[]): void {
  const affected  = depMap.get(mod) || new Map() as AffectedMap;
  if (!depMap.has(mod)) {
    depMap.set(mod, affected);
  }
  affected.set(member, entry);
}

/**
 * Get a file's dependency map
 * @param filePath {string} Absolute file path
 * @param options {object}
 * @prop options.loader {Loader} Read file content from absolute file path. Uses fs.readFileSync as utf8 by default.
 * @prop options.parserOptions {@babel/parser.options} Allow customize babel parser options while parsing file content to AST.
 * @return {Promise<DependencyMap>}
 */
export default async function fileDepMap(filePath: string, { loader, parserOptions }: Options = {}): Promise<DependencyMap> {
  const _loader = loader || ((_filePath: string) => Promise.resolve(fs.readFileSync(_filePath, 'utf8')));
  let file = await _loader(filePath);
  if (!file) {
    throw new Error(`${filePath} is empty!`)
  }
  const { imports: target, exports: entry, relations } = ecmaStats(file, parserOptions);
  debug('>>> ', filePath);
  debug('imports:', target);
  debug('exports:', entry);
  debug('relations:', relations);
  
  const depMap = new Map() as DependencyMap;
  const targetIndex = {} as { [key: string]: Import };

  target.forEach((ref) => {
    const { name, source, alias } = ref;
    targetIndex[alias] = ref;
    updateDependencyMap(depMap, source, name, [] as Entry[]);
  });

  if (entry.extends) {
    entry.extends.forEach((source) => {
      updateDependencyMap(depMap, source, MODULE_ALL, [{
        name: MODULE_ALL,
        source: filePath
      }]);
    })
  }

  debug('targetIndex:', targetIndex);
  debug('depMap initialized:', depMap);

  entry.members.forEach(({ name, alias }) => {
    const dependents = new Set() as Set<Import>;
    let dependsOn = relations[name];
    const toClear = new Set() as Set<string>;
    if (dependsOn) {
      dependsOn.forEach(dependent => {
        const importRef = targetIndex[dependent];
        if (importRef) {
          dependents.add(importRef);
        } else if (relations[dependent]) {
          dependsOn = dependsOn.concat(relations[dependent]);
          toClear.add(dependent);
        }
      });
      /** @todo verify whether this helps for performance */
      relations[name] = dependsOn.filter(name => !toClear.has(name));
    }
    dependents.forEach(importRef => {
      debug('importRef > ', importRef);
      const affectedMap = depMap.get(importRef.source);
      const entries = affectedMap && affectedMap.get(importRef.name);
      if (entries) {
        entries.push({
          name: alias,
          source: filePath
        });
      }
    });
  });

  return depMap;
}
