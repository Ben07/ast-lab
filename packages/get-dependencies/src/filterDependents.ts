import path from 'path';
import fs from 'fs';
import { DEFAULT_EXTENSIONS } from './constants';
import getEs6Dependents from './es6Detect/getEs6Dependents';
import resolveModulePath from './resolveModulePath';
import markDependents from './markDependents';
import hasExt from './hasExt';
import { PathNode,  Exports, Options, Visited, VisitedNode } from './types';

function skipFile(file: string, extensions: string[] | void){
  return extensions && !hasExt(file, extensions) && hasExt(file);
}

function addVisitedNode(visited: Visited, node: PathNode) {
  const { importModule } = node;
  if (visited.has(importModule)) {
    (visited.get(importModule) as VisitedNode).push(node);
    return;
  } else {
    visited.set(importModule, [node]);
  }
}

async function visitPath(visited: Visited, node: PathNode, options: Options) {
  const {
    loader,
    extensions
  } = options;
  const { importModule: source } = node;
  const basePath = path.dirname(source)
  const resolve = async (mod: string) => {
    try {
      return await resolveModulePath(mod, basePath, options);
    } catch (err) {
      console.info('Visiting node: ', node);
      console.log(err);
    }
  };
  const deps = await getEs6Dependents(source, {
    inDetail: false,
    async loader(file: string) {
      if (skipFile(file, extensions)) { return ''; }
      const realPath = await resolve(file);
      return realPath ? loader ? loader(realPath) : fs.readFileSync(realPath, 'utf8') : '';
    }
  });
  addVisitedNode(visited, node);

  let queue = [] as Promise<any>[];
  deps.forEach((i2e, mod) => {
    if (skipFile(mod, extensions)) { 
      return;
    }
    queue.push((async () => {
      const importModule = await resolve(mod);
      if (!importModule) { return; }
      const visitedPaths = visited.get(importModule);
      if (visitedPaths && visitedPaths.find(node => node.source === source)) {
        return;
      }
      const nextNode: PathNode = {
        source,
        importModule,
        i2e,
        prev: node,
      };
      await visitPath(visited, nextNode, options);
    })())
  });
  await Promise.all(queue);
}

export default async function filterDependents(sources: string[], targets: Exports, options: Options = {}): Promise<string[]> {
  // console.log('======================ï£¿=========');
  const visited: Visited = new Map();
  let resolvedSourcePaths = [] as string[];
  options = {
    extensions: DEFAULT_EXTENSIONS,
    ...options
  };
  for(let i = 0; i < sources.length; i++) {
    const sourcePath = sources[i];
    if (!sourcePath || !hasExt(sourcePath) || !path.isAbsolute(sourcePath)) {
      continue;
    }
    resolvedSourcePaths.push(sourcePath);
    if (!visited.has(sourcePath)) {
      const rootNode = {
        source: null,
        importModule: sourcePath,
        i2e: null,
        prev: null
      };
      await visitPath(visited, rootNode, options);
    }
  }
  const marked = new Map(Array.from(targets));
  markDependents(marked, visited);
  return resolvedSourcePaths.filter(s => marked.has(s));
}
