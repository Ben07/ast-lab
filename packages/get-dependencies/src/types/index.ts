import { ALL_EXPORTS } from '../constants';

type Module = string;
type Imported = string;
type ImportedAlias = string;
export type Exported = string;

export type ModuleExported = Set<Exported> | null | typeof ALL_EXPORTS;
export type ModuleImported = Map<Imported, {
  alias: ImportedAlias | null,
  affectedExports: ModuleExported
}>;

export type Dependents = Map<Module, ModuleImported>;
export type Exports = Map<Module, ModuleExported>;

export type ModuleDirectory = string | ReadonlyArray<string>;

export type Alias = {
  [key: string]: string
};

export type Loader = (filename: string) => string;

export type PathNode = {
  source: string | null,
  importModule: string,
  i2e: ModuleImported | null,
  prev: PathNode | null
};

export type Options = {
  loader?: Loader,
  moduleDirectory?: ModuleDirectory,
  extensions?: string[],
  alias?: Alias
};
