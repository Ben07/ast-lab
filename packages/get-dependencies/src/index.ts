import filterDependents from './filterDependents';
import getExports from './es6Detect/getExports';
import hasExt from './hasExt';
export * from './types';

export {
  filterDependents,
  getExports,
  hasExt
};