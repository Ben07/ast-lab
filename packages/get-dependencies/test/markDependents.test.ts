// @ts-ignore
import { enumerateArgsTestFunction, configArgs } from 'lazy-jest';
import _markDependents from '../src/markDependents';
import { Exports, PathNode } from '../src/types';
import { ALL_EXPORTS, ALL_MODULES } from '../src/constants';

function makeSimplePath(paths: string[]) {
  return paths.reduce((ret, p, i) => {
    if (i === paths.length - 1) {
      return ret;
    }
    return {
      source: p,
      importModule: paths[i + 1],
      i2e: new Map([
        [ALL_MODULES, {
          alias: 'ImportAlias',
          affectedExports: ALL_EXPORTS
        }]
      ]),
      prev: ret
    };
  }, {} as PathNode);
}
function markDependents(marked: Exports, paths: string[]) {
  return _markDependents(marked, makeSimplePath(paths));
}

describe('markDependents()', () => {
  const rootNode = {
    source: 'root',
    importModule: 'm2',
    i2e: null,
    prev: null
  };
  enumerateArgsTestFunction(
    _markDependents,
    configArgs()
    .arg('marked', [new Map([
      ['m2', true]
    ])])
    .arg('pathNode', [
      {
        source: 'import-default',
        importModule: 'm2',
        i2e: new Map([
          ['default', {
            alias: 'alias',
            affectedExports: ALL_EXPORTS
          }]
        ]),
        prev: rootNode
      }
    ]),
    'should not mark if end node is not marked'
  );
  enumerateArgsTestFunction(
    markDependents,
    configArgs()
    .arg('marked', [new Map([
      ['m2', true]
    ])])
    .arg('pathNode', [
      ['m1', 'm2', 'm3', 'm4'],
      ['m1', 'm2', 'm3']
    ]),
    'should not mark if end node is not marked'
  );
  enumerateArgsTestFunction(
    markDependents,
    configArgs()
    .arg('marked', [new Map([
      ['m4', true]
    ])])
    .arg('pathNode', [
      ['m1', 'm2', 'm3', 'm4'],
      ['b1', 'b2', 'm2'],
    ]),
    'should mark all paths given'
  );
});
