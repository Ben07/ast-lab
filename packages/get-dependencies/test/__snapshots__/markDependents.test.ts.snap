// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`markDependents() should mark all paths given - (["marked", "pathNode"]) - [{"m4":true},["b1","b2","m4"]] 1`] = `
Map {
  "m4" => true,
  "b2" => true,
  "b1" => true,
}
`;

exports[`markDependents() should mark all paths given - (["marked", "pathNode"]) - [{"m4":true},["m1","m2","m3","m4"]] 1`] = `
Map {
  "m4" => true,
  "m3" => true,
  "m2" => true,
  "m1" => true,
}
`;

exports[`markDependents() should not mark if end node is not marked - (["marked", "pathNode"]) - [{"m2":true},["m1","m2","m3","m4"]] 1`] = `
Map {
  "m2" => true,
  "m1" => true,
}
`;

exports[`markDependents() should not mark if end node is not marked - (["marked", "pathNode"]) - [{"m2":true},["m1","m2","m3"]] 1`] = `
Map {
  "m2" => true,
  "m1" => true,
}
`;

exports[`markDependents() should not mark if end node is not marked 1`] = `
Map {
  "m2" => true,
}
`;
